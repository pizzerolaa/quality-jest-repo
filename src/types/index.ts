export enum TaskStatus {
    TODO = 'TODO',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETED = 'COMPLETED',
}

export enum UserRole {
    WORKER = 'Worker',
    TEAM_LEAD = 'Team Leader',
}

export interface User {
    id: string,
    name: string,
    email: string,
    password: string,
    role: UserRole,
    teamId: string,
}

export interface Task {
    id: string;
    title: string;
    description: string;
    assignedTo: string; // userId
    developerId: string;
    developerName: string;
    status: TaskStatus;
    storyPoints: number;
    estimatedHours: number;
    actualHours: number;
    sprintId: string;
    createdAt: string;
    updatedAt: string;
    completedAt: string | null;
}

export interface Sprint {
    id: string;
    name: string;
    startDate: string;
    endDate: string;
    team: string;
}

export interface Team {
    id: string;
    name: string;
    members: string[]; // array of userIds
    leaderId: string;
}