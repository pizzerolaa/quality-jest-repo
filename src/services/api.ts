import { Task, TaskStatus, User } from '../types';
import { mockTasks, mockUsers, /*mockSprints*/ mockTeams } from '../mocks/data';

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const api = {
  // simulaciones
  getCurrentUser: async (): Promise<User> => {
    await delay(300);
    return mockUsers[0]; //asume el 1er user es el actual
  },

  getAllTasks: async (): Promise<Task[]> => {
    await delay(500);
    return [...mockTasks];
  },

  getTasksBySprint: async (sprintId: string): Promise<Task[]> => {
    await delay(500);
    return mockTasks.filter((task) => task.sprintId === sprintId);
  },

  getTasksByUser: async (userId: string): Promise<Task[]> => {
    await delay(500);
    return mockTasks.filter((task) => task.assignedTo === userId);
  },

  updateTask: async (taskId: string, updates: Partial<Task>): Promise<Task> => {
    await delay(400);
    const taskIndex = mockTasks.findIndex((task) => task.id === taskId);
    if (taskIndex === -1) throw new Error('Task not found');
    const updatedTask = {
      ...mockTasks[taskIndex],
      ...updates,
      updatedAt: new Date().toISOString(),
    };
    mockTasks[taskIndex] = updatedTask;
    return updatedTask;
  },

  completeTask: async (taskId: string, actualHours: number): Promise<Task> => {
    await delay(400);
    return api.updateTask(taskId, {
      status: TaskStatus.COMPLETED,
      completedAt: new Date().toISOString(),
      actualHours,
    });
  },

  getTODOTasksByTeam: async (teamId: string): Promise<Task[]> => {
    await delay(500);
    const teamMembers = mockTeams.find((team) => team.id === teamId)?.members || [];
    return mockTasks.filter(
      (task) => teamMembers.includes(task.assignedTo) && task.status != TaskStatus.COMPLETED
    );
  },
};
