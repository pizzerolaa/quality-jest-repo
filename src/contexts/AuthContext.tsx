import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { User } from '../types';
import { api } from '../services/api';

interface AuthContextType {
    currentUser: User | null;
    loading: boolean;
    error: string | null;
}

const AuthContext = createContext<AuthContextType>({
    currentUser: null,
    loading: true,
    error: null,
});

export const useAuth = () => useContext(AuthContext);

export const AuthProvider: React.FC<{children: ReactNode}> = ({children}) => {
    const [currentUser, setCurrentUser] = useState<User | null>(null);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);

    useEffect(() => {
        const fetchCurrentUser = async () => {
            try {
                const user = await api.getCurrentUser();
                setCurrentUser(user);
            } catch (err) {
                setError('Failed to fetch current user');
                console.error(err);
            } finally {
                setLoading(false);
            }
        };

        fetchCurrentUser();
    }, []);

    return (
        <AuthContext.Provider value={{currentUser, loading, error}}>
            {children}
        </AuthContext.Provider>
    );

    // const value = useMemo(() => ({
    //     currentUser,
    //     loading,
    //     error
    // }), [currentUser, loading, error]);

    // return (
    //     <AuthContext.Provider value={value}>
    //         {children}
    //     </AuthContext.Provider>
    // );
    
};