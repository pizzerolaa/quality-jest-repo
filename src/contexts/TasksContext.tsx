import React, { createContext, useContext, useState, useEffect, ReactNode} from 'react';
import { Task } from '../types';
import { api } from '../services/api';
import { useAuth } from './AuthContext';

interface TasksContextType {
    tasks: Task[];
    loading: boolean;
    error: string | null;
    updateTask: (taskId: string, updates: Partial<Task>) => Promise<void>;
    completeTask: (taskId: string, actualHours: number) => Promise<void>;
    refreshTasks: () => Promise<void>;
}

const TasksContext = createContext<TasksContextType>({
    tasks: [],
    loading: false,
    error: null,
    updateTask: async () => {},
    completeTask: async () => {},
    refreshTasks: async () => {},
});

export const useTasks = () => useContext(TasksContext);

export const TasksProvider: React.FC<{children: ReactNode}> = ({children}) => {
    const [tasks, setTasks] = useState<Task[]>([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);
    const { currentUser } = useAuth();

    const loadTasks = async () => {
        if(!currentUser) return;
        setLoading(true);
        setError(null);
        try {
            let fetchedTasks;
            if(currentUser.role === 'Team Leader') {
                fetchedTasks = await api.getAllTasks();
            } else {
                fetchedTasks = await api.getTasksByUser(currentUser.id);
            }
            setTasks(fetchedTasks);
        } catch (err) {
            setError('Error loading tasks');
            console.error(err);
        } finally {
            setLoading(false);
        }
    };

    useEffect(() => {
        if (currentUser) {
            loadTasks();
        }
    }, [currentUser]);

    const updateTask = async (taskId: string, updates: Partial<Task>) => {
        try {
            await api.updateTask(taskId, updates);
            await loadTasks();
        } catch (err) {
            setError('Error updating task');
            console.error(err);
        }
    };

    const completeTask = async (taskId: string, actualHours: number) => {
        try {
            await api.completeTask(taskId, actualHours);
            await loadTasks();
        } catch (err) {
            setError('Error completing task');
            console.error(err);
        }
    };

    const refreshTasks = async () => {
        await loadTasks();
    };

    return (
        <TasksContext.Provider value={{ tasks, loading, error, updateTask, completeTask, refreshTasks }}>
            {children}
        </TasksContext.Provider>
    );
};
