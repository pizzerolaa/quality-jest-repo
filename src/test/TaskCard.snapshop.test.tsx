// src/__tests__/TaskCard.snapshot.test.tsx
import React from 'react';
import { render } from '@testing-library/react';
import TaskCard from '../components/tasks/TaskCards';
import { TasksProvider } from '../contexts/TasksContext';
import { AuthProvider } from '../contexts/AuthContext';
import { TaskStatus } from '../types';
import { api } from '../services/api';

// Mock the api service
jest.mock('../services/api');

describe('TaskCard Snapshot', () => {
  const mockTask = {
    id: 'task1',
    title: 'Test Task',
    description: 'Description for test task',
    status: TaskStatus.TODO,
    assignedTo: 'A01643639',
    developerId: 'A01643639',
    developerName: 'Sadrac Aramburo',
    storyPoints: 5,
    estimatedHours: 8,
    actualHours: 0,
    sprintId: 'sprint1',
    createdAt: '2025-04-01',
    updatedAt: '2025-04-01',
    completedAt: null,
  };

  beforeEach(() => {
    // Set up API mocks
    api.getCurrentUser = jest.fn().mockResolvedValue({
      id: 'A01643639',
      name: 'Sadrac Aramburo',
      role: 'Worker',
      email: 'sadrac@example.com',
      password: 'password123',
      teamId: 'team1',
    });

    api.getAllTasks = jest.fn().mockResolvedValue([mockTask]);
  });

  test('TaskCard renders correctly for TODO task', () => {
    const { container } = render(
      <AuthProvider>
        <TasksProvider>
          <TaskCard task={mockTask} />
        </TasksProvider>
      </AuthProvider>
    );

    expect(container).toMatchSnapshot();
  });

  test('TaskCard renders correctly for IN_PROGRESS task', () => {
    const inProgressTask = {
      ...mockTask,
      status: TaskStatus.IN_PROGRESS,
    };

    const { container } = render(
      <AuthProvider>
        <TasksProvider>
          <TaskCard task={inProgressTask} />
        </TasksProvider>
      </AuthProvider>
    );

    expect(container).toMatchSnapshot();
  });

  test('TaskCard renders correctly for COMPLETED task', () => {
    const completedTask = {
      ...mockTask,
      status: TaskStatus.COMPLETED,
      completedAt: '2025-04-10',
    };

    const { container } = render(
      <AuthProvider>
        <TasksProvider>
          <TaskCard task={completedTask} />
        </TasksProvider>
      </AuthProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
