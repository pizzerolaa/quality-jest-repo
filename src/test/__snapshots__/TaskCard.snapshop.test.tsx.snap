// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCard Snapshot TaskCard renders correctly for COMPLETED task 1`] = `
<div>
  <div
    class="border rounded-lg p-4 mb-4 shadow-sm bg-green-100 border-green-400"
  >
    <div
      class="flex justify-between items-start"
    >
      <h3
        class="font-bold text-lg"
      >
        Test Task
      </h3>
      <span
        class="text-sm font-medium px-2 py-1 rounded-full bg-gray-200"
      >
        COMPLETED
      </span>
    </div>
    <p
      class="text-gray-600 mt-1"
    >
      Description for test task
    </p>
    <div
      class="mt-3 grid grid-cols-2 gap-2"
    >
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Developer:
        </p>
        <p>
          Sadrac Aramburo
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Story Points:
        </p>
        <p>
          5
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Estimated Hours:
        </p>
        <p>
          8
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Actual Hours:
        </p>
        <p>
          0
        </p>
      </div>
    </div>
    <div
      class="mt-4 flex gap-2"
    />
  </div>
</div>
`;

exports[`TaskCard Snapshot TaskCard renders correctly for IN_PROGRESS task 1`] = `
<div>
  <div
    class="border rounded-lg p-4 mb-4 shadow-sm bg-blue-100 border-blue-400"
  >
    <div
      class="flex justify-between items-start"
    >
      <h3
        class="font-bold text-lg"
      >
        Test Task
      </h3>
      <span
        class="text-sm font-medium px-2 py-1 rounded-full bg-gray-200"
      >
        IN_PROGRESS
      </span>
    </div>
    <p
      class="text-gray-600 mt-1"
    >
      Description for test task
    </p>
    <div
      class="mt-3 grid grid-cols-2 gap-2"
    >
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Developer:
        </p>
        <p>
          Sadrac Aramburo
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Story Points:
        </p>
        <p>
          5
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Estimated Hours:
        </p>
        <p>
          8
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Actual Hours:
        </p>
        <p>
          0
        </p>
      </div>
    </div>
    <div
      class="mt-4 flex gap-2"
    >
      <button
        class="bg-blue-500 text-white px-3 py-1 rounded hover:bg-blue-600"
      >
        Edit
      </button>
      <div
        class="flex items-center"
      >
        <input
          class="border rounded w-16 px-2 py-1 mr-2"
          placeholder="Hours"
          type="number"
          value="0"
        />
        <button
          class="bg-green-500 text-white px-3 py-1 rounded hover:bg-green-600"
        >
          Complete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskCard Snapshot TaskCard renders correctly for TODO task 1`] = `
<div>
  <div
    class="border rounded-lg p-4 mb-4 shadow-sm bg-yellow-100 border-yellow-400"
  >
    <div
      class="flex justify-between items-start"
    >
      <h3
        class="font-bold text-lg"
      >
        Test Task
      </h3>
      <span
        class="text-sm font-medium px-2 py-1 rounded-full bg-gray-200"
      >
        TODO
      </span>
    </div>
    <p
      class="text-gray-600 mt-1"
    >
      Description for test task
    </p>
    <div
      class="mt-3 grid grid-cols-2 gap-2"
    >
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Developer:
        </p>
        <p>
          Sadrac Aramburo
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Story Points:
        </p>
        <p>
          5
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Estimated Hours:
        </p>
        <p>
          8
        </p>
      </div>
      <div>
        <p
          class="text-sm text-gray-500"
        >
          Actual Hours:
        </p>
        <p>
          0
        </p>
      </div>
    </div>
    <div
      class="mt-4 flex gap-2"
    >
      <button
        class="bg-blue-500 text-white px-3 py-1 rounded hover:bg-blue-600"
      >
        Edit
      </button>
      <div
        class="flex items-center"
      >
        <input
          class="border rounded w-16 px-2 py-1 mr-2"
          placeholder="Hours"
          type="number"
          value="0"
        />
        <button
          class="bg-green-500 text-white px-3 py-1 rounded hover:bg-green-600"
        >
          Complete
        </button>
      </div>
    </div>
  </div>
</div>
`;
